test/e2e/antreapolicy_test.go

```diff
@@ -1920,6 +1920,72 @@ func testANPBasic(t *testing.T) {
 	executeTests(t, testCase)
 }
 
+// testANPMultipleAppliedTo tests traffic from X/B to Y/A on port 80 will be dropped, after applying Antrea
+// NetworkPolicy that applies to multiple AppliedTos, one of which doesn't select any Pod. It also ensures the Policy is
+// updated correctly when one of its AppliedToGroup starts and stops selecting Pods.
+func testANPMultipleAppliedTo(t *testing.T, singleRule bool) {
+	tempLabel := randName("temp-")
+	builder := &AntreaNetworkPolicySpecBuilder{}
+	builder = builder.SetName("y", "np-multiple-appliedto").SetPriority(1.0)
+	// Make it apply to an extra dummy AppliedTo to ensure it handles multiple AppliedToGroups correctly.
+	// See https://github.com/vmware-tanzu/antrea/issues/2083.
+	if singleRule {
+		builder.SetAppliedToGroup([]ANPAppliedToSpec{{PodSelector: map[string]string{"pod": "a"}}, {PodSelector: map[string]string{tempLabel: ""}}})
+		builder.AddIngress(v1.ProtocolTCP, &p80, nil, nil, nil, map[string]string{"pod": "b"}, map[string]string{"ns": "x"},
+			nil, nil, nil, crdv1alpha1.RuleActionDrop, "")
+	} else {
+		builder.AddIngress(v1.ProtocolTCP, &p80, nil, nil, nil, map[string]string{"pod": "b"}, map[string]string{"ns": "x"},
+			nil, nil, []ANPAppliedToSpec{{PodSelector: map[string]string{"pod": "a"}}}, crdv1alpha1.RuleActionDrop, "")
+		builder.AddIngress(v1.ProtocolTCP, &p80, nil, nil, nil, map[string]string{"pod": "b"}, map[string]string{"ns": "x"},
+			nil, nil, []ANPAppliedToSpec{{PodSelector: map[string]string{tempLabel: ""}}}, crdv1alpha1.RuleActionDrop, "")
+	}
+
+	reachability := NewReachability(allPods, Connected)
+	reachability.Expect(Pod("x/b"), Pod("y/a"), Dropped)
+
+	_, err := k8sUtils.CreateOrUpdateANP(builder.Get())
+	failOnError(err, t)
+	time.Sleep(networkPolicyDelay)
+	k8sUtils.Validate(allPods, reachability, 80, v1.ProtocolTCP)
+	reachability.PrintSummary(true, true, true)
+	_, wrong, _ := reachability.Summary()
+	if wrong != 0 {
+		t.Errorf("failure -- %d wrong results", wrong)
+	}
+
+	t.Logf("Making the Policy apply to y/c by labeling it with the temporary label that matches the dummy AppliedTo")
+	podYC, _ := k8sUtils.GetPodByLabel("y", "c")
+	podYC.Labels[tempLabel] = ""
+	podYC, err = k8sUtils.clientset.CoreV1().Pods(podYC.Namespace).Update(context.TODO(), podYC, metav1.UpdateOptions{})
+	assert.NoError(t, err)
+	reachability = NewReachability(allPods, Connected)
+	reachability.Expect(Pod("x/b"), Pod("y/a"), Dropped)
+	reachability.Expect(Pod("x/b"), Pod("y/c"), Dropped)
+	time.Sleep(networkPolicyDelay)
+	k8sUtils.Validate(allPods, reachability, 80, v1.ProtocolTCP)
+	reachability.PrintSummary(true, true, true)
+	_, wrong, _ = reachability.Summary()
+	if wrong != 0 {
+		t.Errorf("failure -- %d wrong results", wrong)
+	}
+
+	t.Logf("Making the Policy not apply to y/c by removing the temporary label")
```

Comment:
nit: we can set these as separate testSteps by not providing any new resources in each testStep, but simply twerk the Pod labels in between testSteps. This way we can get rid of the repeated Validate() and Summary() calls?

Reply:
I tried to do it but didn't find how to insert operations between testSteps with existing code. I tried to add a function field to testStep struct, however most of its consumer doesn't specify fields when constructing instances, so I will have to update all test cases which I want to avoid in this PR.

Reply:
Make sense. I can take that as a follow-up.